### PR TITLE
Fix DynamicResource in Style resources

### LINKED ISSUE #3590: Resources in Styles don't work properly.
From @MarchingCube:

> In Styles.Resources - DynamicResource does nothing, works if I set color to a constant like Red. Next StaticResource causes stack overflow.
> In Style.Resources - DynamicResource does not work either

```
<Styles xmlns="https://github.com/avaloniaui" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">

  <Styles.Resources>
        <SolidColorBrush x:Key="Test" Color="{DynamicResource ThemeControlHighColor}" />
        <!--<SolidColorBrush x:Key="Test" Color="{StaticResource ThemeControlHighColor}" />-->
  </Styles.Resources>

  <Style Selector="Button">
    <Style.Resources>
      <!--<SolidColorBrush x:Key="Test" Color="{DynamicResource ThemeControlHighColor}" />-->
    </Style.Resources>
        <Setter Property="Background" Value="{DynamicResource Test}"/>
    <Setter Property="BorderBrush" Value="{DynamicResource ThemeBorderLowBrush}"/>
    <Setter Property="BorderThickness" Value="{DynamicResource ThemeBorderThickness}"/>
    <Setter Property="Foreground" Value="{DynamicResource ThemeForegroundBrush}"/>
    <Setter Property="HorizontalContentAlignment" Value="Center"/>
    <Setter Property="VerticalContentAlignment" Value="Center"/>
    <Setter Property="Padding" Value="4"/>
    <Setter Property="Template">
      <ControlTemplate>
        <ContentPresenter Name="PART_ContentPresenter"
                          Background="{TemplateBinding Background}"
                          BorderBrush="{TemplateBinding BorderBrush}"
                          BorderThickness="{TemplateBinding BorderThickness}"
                          ContentTemplate="{TemplateBinding ContentTemplate}"
                          Content="{TemplateBinding Content}"
                          Padding="{TemplateBinding Padding}"
                          TextBlock.Foreground="{TemplateBinding Foreground}"
                          HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                          VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}"/>
      </ControlTemplate>
    </Setter>
  </Style>
  <Style Selector="Button:pointerover /template/ ContentPresenter">
    <Setter Property="BorderBrush" Value="{DynamicResource ThemeBorderMidBrush}"/>
  </Style>
  <Style Selector="Button:pressed  /template/ ContentPresenter">
    <Setter Property="Background" Value="{DynamicResource ThemeControlHighBrush}"/>
  </Style>
  <Style Selector="Button:disabled">
    <Setter Property="Opacity" Value="{DynamicResource ThemeDisabledOpacity}"/>
  </Style>
</Styles>
```

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

As #3590 describes, placing a resource that itself uses a `{DynamicResouce}` in a `Style.Resources` or `Styles.Resources` caused the dynamic resource to not take effect.

This was because the parent `Styles` object wasn't informing added child styles when resources change, and of course resources can change for a `Style` when is added as a child of a `Styles` object.

Bonus change: make `StyledElement.Styles` readonly. It was already readonly in Application and I can't see a reason it should be writable here. I think this is a remnant of very early code.

## How was the solution implemented (if it's not obvious)?

Call `ISetResourceParent.ParentResourcesChanged` on a style when it is added to a `Styles` collection and when the parent receives a notification that resources have changed. To make this work I had to prevent re-entrancy when notifying of resource changes to prevent stack overflows.

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues

Fixes #3590

### CODE REVIEW

File: src/Avalonia.Styling/StyledElement.cs [hunk 1]
```diff
@@ -235,6 +236,7 @@ public Styles Styles
                     }
 
                     _styles.ResourcesChanged += ThisResourcesChanged;
+                    NotifyResourcesChanged(new ResourcesChangedEventArgs());
```
Comment (MarchingCube) [inline src/Avalonia.Styling/StyledElement.cs]: What do you think about caching args instance in a static and reusing that? Would save us several thousands allocations when starting the app and loading more complex controls.

File: src/Avalonia.Styling/Styling/Styles.cs [hunk 2]
```diff
@@ -38,7 +39,7 @@ public Styles()
                         ResourcesChanged?.Invoke(this, new ResourcesChangedEventArgs());
                     }
 
-                    x.ResourcesChanged += SubResourceChanged;
+                    x.ResourcesChanged += NotifyResourcesChanged;
```
Comment (MarchingCube) [inline src/Avalonia.Styling/Styling/Styles.cs]: Not something that we want to fix within this PR but I wanted to ask if we have any other alternatives to using events for this? We create so many delegates for this that in calendar benchmark it was like the 6th top allocation.

image

image

File: src/Avalonia.Styling/StyledElement.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Styling/StyledElement.cs]: Yeah we should do that. My vague plan was to indicate which resources changed in the event args, which would mean we can't do that, but for now it's worth doing.

File: src/Avalonia.Styling/Styling/Styles.cs (hunk 2 above)
Comment (grokys) [reply src/Avalonia.Styling/Styling/Styles.cs]: Yeah we could probably remove the usage of events for this.

**Review by MarchingCube — APPROVED** [review]
LGTM. Verified that locally in our repro case and it fixed all resource issues. There is still one issue with converters that we discussed on slack but I think we can do that separately.

**Pull request merged**